numba/np/arraymath.py

```diff
@@ -669,6 +670,136 @@ def _trivial_isnan(x):
+
+
+@overload(np.iscomplexobj)
+def iscomplexobj(x):
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/lib/type_check.py#L282-L320
+    val = x.key if isinstance(x, types.Optional) else x
+    dt = determine_dtype(val)
+    iscmplx = np.issubdtype(dt, np.complexfloating)
+    return lambda x: False if x is None else iscmplx
+
+
+@overload(np.isrealobj)
+def isrealobj(x):
+    # Return True if x is a not complex type or an array of complex numbers.
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/ccfbcc1cd9a4035a467f2e982a565ab27de25b6b/numpy/lib/type_check.py#L290-L322
+    def impl(x):
+        return not np.iscomplexobj(x)
+    return impl
+
+
+@overload(np.isscalar)
+def np_isscalar(num):
+    res = isinstance(num, (types.Number, types.UnicodeType, types.Boolean))
+
+    def impl(num):
+        return res
+    return impl
+
+
+@register_jitable
+def _within_tol(a, b, rtol, atol):
+    return np.less_equal(np.abs(a - b), atol + rtol * np.abs(b))
+
+
+@overload(np.isclose)
+def np_isclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
+    # Based on NumPy impl.
+    # https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/core/numeric.py#L2180-L2292
+
+    if not(type_can_asarray(a) and type_can_asarray(b)):
+        raise TypingError("`np.isclose` is not supported for the input types.")
+
+    def impl(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
+        xfin = np.asarray(np.isfinite(a))
+        yfin = np.asarray(np.isfinite(b))
+        if np.all(xfin) and np.all(yfin):
+            return _within_tol(a, b, rtol, atol)
+        else:
+            r = _within_tol(a, b, rtol, atol)
+            if equal_nan:
+                return r | (~xfin & ~yfin)
+            else:
+                return r
+    return impl
+
+
+def is_np_inf_impl(x, out, fn):
+
+    if not type_can_asarray(x):
+        raise TypingError("First argument must be array-like")
+
+    if is_nonelike(out):
+        if isinstance(x, (types.Array, types.Sequence, types.Tuple)):
+            def impl(x, out=None):
+                x = np.asarray(x)
+                out = np.zeros(x.shape, dtype=types.boolean)
+                np.logical_and(np.isinf(x), fn(np.signbit(x)), out)
+                return out
+        else:
+            def impl(x, out=None):
+                out = np.isinf(x) and fn(np.signbit(x))
+                return out
+    else:
+        def impl(x, out=None):
+            x = np.asarray(x)
+            out = np.asarray(out)
+            # https://github.com/numpy/numpy/blob/b7c27bd2a3817f59c84b004b87bba5db57d9a9b0/numpy/doc/broadcasting.py#L52
+            if (x.size == 1 and x.ndim < out.ndim) or (x.shape == out.shape):
+                np.logical_and(np.isinf(x), fn(np.signbit(x)), out)
+                return out
+
+            raise ValueError('"x" and "out" have different shapes')
+
```

Comment:
I'm not really sure what's going on here, but this `if` seems wrong to me. Can you just make this:

```suggestion
            return np.logical_and(np.isinf(x), fn(np.signbit(x)), out)
```

`np.isinf(np.zeros((3, 1)), out=np.zeros((3, 2)))` works just fine for me, but looks like is forbidden here.

Reply:
Thanks for the tip!

Reply:
This part was necessary due to inputs like `np.isneginf(np.array([np.NINF, np.NINF]), np.zeros((3,)))`

Reply:
Why, what output does this code give with my suggestion?

Reply:
Input: `np.isneginf(np.array([np.NINF, np.NINF]), np.zeros((3,)))`

NumPy output: `ValueError: operands could not be broadcast together with shapes (2,) (2,) (3,)`

Numba output: `[1. 1. 0.]`

Reply:
Then `np.logical_and` is broken in numba. Working around arbitrary bugs at every caller isn't scalable - I'd leave it broken, and file an issue to fix `np.logical_and`.

Simple failing test:
```python
In [21]: @numba.njit
    ...: def foo():
    ...:     return np.logical_and(np.ones(2), np.ones(2), np.zeros(3))
    ...:

In [22]: foo()
Out[22]: array([1., 1., 1.])
```